index.js

```diff
@@ -2,24 +2,207 @@
+      // Output for checking:
+      // console.log(`--- \nIDP realm: ${idpRealm.id}`);
+      // console.info(`IDP setup: ${idpAlias}`);
+      // console.log('Clients:');
+      // console.info(idpClientIds);
+
+      // get all clients for APP realms:
+      const appRealmClients = idpClientIds.filter(c => c.includes('https'));
+      clientList.push({
+        realm: r.ALIAS,
+        clients: appRealmClients,
+      });
+      return clientList;
+    }, Promise.resolve([]));
+
+    if (!MODE.EXECUTE_CHANGE) {
+      console.log('------------- all clients from IDP realms:');
+      console.log(allIdpClients);
+    }
+
+    // ------------------------------------------------------------------------------------------------------
+    // 2. get all APP realms (other than IDP realms), and the IDPs in each realm:
+
+    const allRealms = await kcAdminClient.realms.find();
+    const idpRealmNames = IDP_TERMS.map(r => r.REALM);
+    const appRealms = allRealms.reduce((acc, r) => {
+      if (!idpRealmNames.includes(r.realm)) {
+        const idpAlias = r.identityProviders? r.identityProviders.map(idp => idp.alias) : [];
+
+        acc.push({
+          realm: r.realm,
+          idps: idpAlias,
+        });
+      }
+      return acc;
+    }, []);
+
+    if (!MODE.EXECUTE_CHANGE) {
+      console.log('------------- all APP realms with IDP list:');
+      console.log(appRealms);
+    }
+
+    // ------------------------------------------------------------------------------------------------------
+    // 3. compare if IDP clients and app realms matches
+
+    // - Set statistic:
+    let clientsCount = 0;
+    let idpsCount = 0;
+    let clientsCountArray = [];
+    let idpsCountArray = [];
+
+    // Form a list of all clients for each IDP realm:
+    allIdpClients.forEach(i => {
+      i.clients.forEach(c => {
+        clientsCount++;
+        clientsCountArray.push(`${c}/${i.realm}`);
+      })
+    })
+    // Form a list of all app realms and corresponding IDP usage:
+    appRealms.forEach(r => {
+      r.idps.forEach(i => {
+        idpsCount++;
+        idpsCountArray.push(`https://${oldRoute}/auth/realms/${r.realm}/${i}`);
+      })
+    })
+
+    // - Display extra items that should not be updated:
+    console.log(`idpsCount: ${idpsCount}, clientsCount: ${clientsCount}`);
+    const extraClients = clientsCountArray.filter(x => !idpsCountArray.includes(x));
+    const extraIDPs = idpsCountArray.filter(x => !clientsCountArray.includes(x));
+    console.log('extraClients');
+    console.log(extraClients);
+    console.log('extraIDPs');
+    console.log(extraIDPs);
+
+    // ------------------------------------------------------------------------------------------------------
+    // 4. update url in IDP realms' clients for all APP realms
+
+    console.log('------------- Update url in IDP realms clients for all APP realms:');
+    await allIdpClients.forEach(async r => {
```

Comment:
can run async functions in`forEach` !!

Reply:
you need to run map, since those promises need to resolve in order to gaurantee you're doing things in order.